pkg/audit/audit_client.go

```diff
@@ -0,0 +1,160 @@
+	if !c.enabled {
+		return
+	}
+	auditObject := RootEvent{
+		ResourceType: "AuditEvent",
+		Event: &Event{
+			Type: &Coding{
+				System: "http://hl7.org/fhir/ValueSet/audit-event-type", Version: "1", Code: "110114", Display: "User Authentication"},
+			Action:      "E",
+			DateTime:    time.Now().UTC().Format(time.RFC3339),
+			Outcome:     outcomeCode,
+			OutcomeDesc: outcomeDesc},
+
+		Participant: []*Participant{
+			{AltID: ss.User, UserID: UserID{Value: ss.Email}, Name: ss.PreferredUsername, Requestor: true}},
+		Source: Source{
+			Identifier: Identifier{
+				Type: &Coding{
+					System:  "http://hl7.org/fhir/ValueSet/identifier-type",
+					Code:    "4",
+					Display: "Application Server",
+				},
+				Value: ss.Email,
+			},
+			Type: []*Coding{{System: "http://hl7.org/fhir/security-source-type", Code: "1", Display: "End-user display device, diagnostic device."}},
+			Extension: []*Extension{
+				{
+					URL: appURL,
+					Extension: []*ExtensionContent{
+						{
+							URL:         "applicationName",
+							ValueString: c.opts.ProductName,
+						},
+						{
+							URL:         "applicationVersion",
+							ValueString: "1",
+						},
+						{
+							URL:         "serverName",
+							ValueString: "oauth2proxy",
+						},
+						{
+							URL:         "componentName",
+							ValueString: "oauth2proxy",
+						},
+						{
+							URL:         "productKey",
+							ValueString: c.opts.ProductKey,
+						},
+						{
+							URL:         "tenant",
+							ValueString: tenantID,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	auditMessage, err := json.Marshal(auditObject)
+	if err != nil {
+		logger.Errorf("%s: could not marshal the audit object: %v", ErrPersitAuditEvent.Error(), err)
+		return
+	}
+	err = c.send(string(auditMessage))
+	if err != nil {
+		logger.Errorf("%s: could not send the audit message to the url '%s': %v", ErrPersitAuditEvent.Error(), c.opts.URL, err)
+		return
+	}
+}
+
+func (c *Client) send(msg string) error {
+	signedDate := time.Now().UTC().Format(time.RFC3339)
+	signature := c.apiSignature.GetSignature(signedDate)
+	resp, err := c.client.R().
+		SetHeader("Content-Type", "application/json").
+		SetHeader("api-version", "2").
+		SetHeader("HSDP-API-Signature", signature).
+		SetHeader("SignedDate", signedDate).
+		SetBody(msg).
+		Post(c.opts.URL)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode() != 201 {
+		log.Println("Not able to send the audit message ", resp)
+		return fmt.Errorf("not able to persist audit, audit server returned %v", resp.StatusCode())
+	}
+	return nil
+}
+
+func (c *ClientOpts) Validate() error {
+	if strings.TrimSpace(c.ProductName) == "" || strings.TrimSpace(c.ProductKey) == "" || strings.TrimSpace(c.SecretKey) == "" || strings.TrimSpace(c.SharedKey) == "" {
+		return errors.New("the audit is enabled and therefore the audit product name, audit key, audit secret key or audit shared key are required (however found empty)")
```

Comment:
You created audit_error.go, why this error is not there?

Reply:
this is more a setup error and the others are actually runtime errors